docs/src/model_setup/boundary_conditions.md

```diff
@@ -487,8 +487,17 @@ Oceananigans.FieldBoundaryConditions, with boundary conditions
 └── immersed: ImmersedBoundaryCondition with west=Nothing, east=Nothing, south=Nothing, north=Nothing, bottom=Value, top=Nothing
 ```
 
-A boundary condition that depends on the fields may be prescribed using the `immersed` keyword argument in [`FieldBoundaryConditions`](@ref).
-We illustrate field-dependent boundary conditions with an example that imposes linear bottom drag on `u` on both immersed facets _and_ the bottom boundary of the underlying grid.
+!!! warning "`ImmersedBoundaryCondition`"
+    Manually building `ImmersedBoundaryCondition` should be done with care.
+    While convenient for some applications, using bespoke `ImmersedBoundaryCondition`s
+    can obscure the physical interpreation of a model setup. As with all model setups, it is important
+    to test important model choices in idealized situations where their impact can be easily discerned.
```

Comment:
I don't know about this. I guess I don't really understand what's bad about `ImmersedBoundaryCondition`. I agree, it's a bit less clear than the alternative, but I don't understand why it merits a warning. 

The same goes for the part about model choices. It's true, but I guess that's true of all other important choices and I don't see why it needs a warning.

Is there any reason why we don't just test `ImmersedBoundaryCondition` more thoroughly and treat it like any other BC in the docs?

Reply:
I think we should include some comment that encourages using ordinary boundary conditions. I'm worried that the name is confusing: users could interpret `ImmersedBoundaryCondition` as being the only way to specify a condition on immersed boundaries.

More generally, I think that we can interpret cases that require this sort of detailed boundary condition construction (which typically would not generalize well) as a situation in which possibly more source code / more features are needed. In other words, we support this because it's not possible to support all boundary conditions models in a nice general fashion within Oceananigans. But I'd hope that through development efforts that lower-level aspects of the code, like `ImmersedBoundaryCondition`, would be used less often than higher-level, more generalizable (and testable) abstractions.

Reply:
The problem with complex implementations in user scripts is that they are harder to test and validate. Generally I think we try to design the code to encourage reproducible and high-quality science. I'm seeing a sort of gap here, where it's possible to "see between the cracks". Something similar goes on with `DiscreteForcing`, though our comment there is a bit more mild "Using discrete forcing functions may require understanding the staggered arrangement of velocity fields and tracers..."

https://clima.github.io/OceananigansDocumentation/stable/model_setup/forcing_functions/#%22Discrete-form%22-forcing-functions

Reply:
> The problem with complex implementations in user scripts is that they are harder to test and validate. Generally I think we try to design the code to encourage reproducible and high-quality science. I'm seeing a sort of gap here, where it's possible to "see between the cracks". Something similar goes on with `DiscreteForcing`, though our comment there is a bit more mild "Using discrete forcing functions may require understanding the staggered arrangement of velocity fields and tracers..."
> 
> https://clima.github.io/OceananigansDocumentation/stable/model_setup/forcing_functions/#%22Discrete-form%22-forcing-functions

I think I understand your point a bit better. I guess part of your argument is that this API is in place only because we haven't been able to implement something clearer/more user-friendly yet?

If so, how about the following as a warning?:

```suggestion
!!! warning "`ImmersedBoundaryCondition`"
    `ImmersedBoundaryCondition` is considered an experimental API and may change in the future.
    Therefore, one should use it only when a finer level of control over the boundary conditions
    at the immersed boundary is required, and the user is familiar with the implementation of boundary 
    conditions on staggered grids. For all other cases , using the `immersed` argument of
    `FieldBoundaryConditions` is preferred.
```

Reply:
I like this version, is a bit more clear IMO